README.md
# Flickr Objects

This gem is an object-oriented wrapper for the [Flickr API](http://flickr.com/api).

- Web page: [http://janko-m.github.com/flickr-objects](http://janko-m.github.com/flickr-objects)
- Source code: [https://github.com/janko-m/flickr-objects](https://github.com/janko-m/flickr-objects)
- API Documentation: [http://rubydoc.info/github/janko-m/flickr-objects/master/frames](http://rubydoc.info/github/janko-m/flickr-objects/master/frames)

The gem has been tested on the following Ruby versions:

- MRI 1.9.3
- MRI 2.0
- MRI 2.1

## Installation and setup

Add it to your `Gemfile`, and run `bundle install`.

```ruby
gem "flickr-objects"
```

Now create an initializer where you set your Flickr credentials.

```ruby
Flickr.configure do |config|
  config.api_key       = "API_KEY"
  config.shared_secret = "SHARED_SECRET"
end
```

If you don't have them yet, you can apply for them
[here](http://www.flickr.com/services/apps/create/apply).

For list of possible configuration options, see
[`Flickr::Configuration`](http://rubydoc.info/github/janko-m/flickr-objects/master/Flickr/Configuration).

## Usage

This gem maps [Flickr's API methods](http://flickr.com/api) to Ruby methods.

```ruby
Flickr.photos.search(...)   # flickr.photos.search
Flickr.person.get_sets(...) # flickr.photosets.getList
```

For the list of all API methods (and their related Flickr's methods), see
[`Flickr::Api`](http://rubydoc.info/github/janko-m/flickr-objects/master/Flickr/Api).
These methods are included to the `Flickr` module.

Example:

```ruby
photos = Flickr.photos.search(user_id: "78733179@N04") #=> [#<Flickr::Object::Photo: ...>, #<Flickr::Object::Photo: ...>, ...]

photo = photos.first
photo.id                 #=> "231233252"
photo.title              #=> "My cat"
photo.visibility.public? #=> true

person = Flickr.people.find("78733179@N04")
set = person.sets.first
set.id           #=> "11243423"
set.photos_count #=> 40
```

Few notes here:

- flickr-objects distinguishes **instance** from **class** API methods. So,
  `Flickr.photos.search` is a **class** API method. And `Flickr.people.get_sets`
  is by its nature an **instance** API method, because we're finding sets
  *from a person*; in that case it's nicer to call `#get_sets` on an instance of
  person.
- Flickr objects can always be instantiated with `Flickr.<objects>.find(id)`
  (in the above example we did `Flickr.people.find(id)`).

## Arguments to API methods

By its nature, API methods always accept a hash of parameters. Using that approach,
this would be the call for "flickr.people.findByEmail":

```ruby
Flickr.people.find_by_email(find_email: "janko.marohnic@gmail.com")
```

But that looks lame. Luckily, in these kind of methods flickr-objects improves
the argument list:

```ruby
Flickr.people.find_by_email("janko.marohnic@gmail.com")
```

These arguments are documented:
[`Flickr::Api::Person#find_by_email`](http://rubydoc.info/github/janko-m/flickr-objects/master/Flickr/Api/Person#find_by_email-instance_method).

## Sizes

```ruby
person = Flickr.person.find("78733179@N04")
photo = person.public_photos(sizes: true).first

photo.small!(320)
photo.source_url #=> "http://farm9.staticflickr.com/8191/8130464513_780e01decd_n.jpg"
photo.width      #=> 320
photo.height     #=> 280

photo.medium!(500)
photo.width      #=> 500
```

It is important here that you pass `sizes: true` to `Flickr::Person#public_photos`.
So, in your (Rails) application, you could use it like this:

```ruby
class PhotosController < ApplicationController
  def index
    person = Flickr.people.find("78733179@N04")
    @photos = person.public_photos(sizes: true).map(&:medium500!)
  end
end
```
```erb
<% @photos.each do |photo| %>
  <%= image_tag photo.source_url, size: "#{photo.width}x#{photo.height}" %>
<% end %>
```

To find out more, see [`Flickr::Object::Photo`](http://rubydoc.info/github/janko-m/flickr-objects/master/Flickr/Object/Photo).

## Authentication

You may need to make authenticated API requests, using an access token.

```ruby
flickr = Flickr.new("ACCESS_TOKEN_KEY", "ACCESS_TOKEN_SECRET")

# It has the same interface as `Flickr`
flickr.test_login #=> {"id" => "78733179@N04", "username" => ...}
flickr.people.find("78733179@N04").get_photos #=> [#<Flickr::Photo ...>, #<Flickr::Photo, ...>, ...]
```

For details on how to authenticate the user, that is, obtain his access token, see
[`Flickr::OAuth`](http://rubydoc.info/github/janko-m/flickr-objects/master/Flickr/OAuth).

If you want, you can also assign the access token globally in your configuration.

```ruby
Flickr.configure do |config|
  config.access_token_key = "ACCESS_TOKEN_KEY"
  config.access_token_secret = "ACCESS_TOKEN_SECRET"
end
```

Naturally, this way you don't need to create an instance like above.

## Upload

```ruby
photo_id = Flickr.upload("/path/to/photo.jpg", title: "Dandelions")
photo = Flickr.photos.find(photo_id).get_info!
photo.title #=> "Dandelions"
```

See [`Flickr.upload`](http://rubydoc.info/github/janko-m/flickr-objects/master/Flickr/Api/General#upload-instance_method).

## Pagination

```ruby
Flickr.configure do |config|
  config.pagination = :will_paginate
end
```
```ruby
@photos = Flickr.photos.search(user_id: "78733179@N04", page: params[:page], per_page: 10)
```
```erb
<%= will_paginate @photos %>
```

Flickr gives you pagination on almost every request that returns a collection of objects.
This gem supports both [WillPaginate](https://github.com/mislav/will_paginate) (`:will_paginate`)
and [Kaminari](https://github.com/amatsuda/kaminari) (`:kaminari`).

## Caching

To enable caching responses, just pass in a cache object (an object that responds
to `#read`, `#write` and `#fetch`) in the configuration.

```ruby
require "active_support/cache"
require "active_support/core_ext/numeric/time"

Flickr.configure do |config|
  config.cache = ActiveSupport::Cache::MemoryStore(expires_in: 1.week)
end
```

```ruby
Flickr.photos.recent # Makes an API call
Flickr.photos.recent # Uses the cache
```

## Few words

Many of the API methods are not covered yet (because they are so many).
I believe I covered all the important ones, but if you wish for me to
cover certain new ones, feel free to contact me via [Twitter](https://twitter.com/jankomarohnic).
I also wouldn't mind getting a pull request ;)

## Social

You can follow me on Twitter, I'm [@jankomarohnic](https://twitter.com/jankomarohnic).

## License

This project is released under the [MIT license](LICENSE).


CHANGELOG.md
# Changelog

## Version 0.6.3

- Add ability to upload any IO-like objects

## Version 0.6.2

- Don't modify the parsed response hash anymore. When Flickr is used with a
  memory cache, the response hash is hashed in memory. Because we were
  modifying this hash, this also modified the cached result, causing errors
  because of missing data.

## Version 0.6.1

- Remove the default timeout options, so the timeout will happen only if the
  user sets one.

- Make `Flickr::Object::Photo#size!` public.

- Allow uploading any kind of IO, not just files.

- Make `Flickr::Object::Photo#available_sizes` return only sizes that are
  actually available.

- Add `Flickr::Object::Set#photos` as an alias to `#get_photos`, to be
  consistent with other objects.

## Version 0.6.0

- Add license methods:

  ```
  flickr.photos.licenses.getInfo
  flickr.photos.licenses.setLicense
  ```

## Version 0.5.2

- Require Flickr errors, so that they're rescuable (thanks to @erik-eide)
- Make `Flickr::ApiError`'s message optional, so that it can be used by RSpec's
  `raise_error` matcher, without having to provide a second argument

## Version 0.5.1

- Generate different size's source URLs from photo's info
  ([docs](https://www.flickr.com/services/api/misc.urls.html)), if sizes are
  missing. That means that for some sizes ("Square 75" to "Large 1024")
  there is no need to pass in the `sizes: true` option.

## Version 0.5.0

- Add support for caching reponses.

## Version 0.4.1

- Flickr's API transferred to SSL, and the calls over http:// no longer work,
  so flickr-objects now makes calls always over SSL.

## Version 0.4.0

- Added `flickr.interestingness.getList`.
- Fixed `sizes: true` not applying to `flickr.photos.getContactsPublicPhotos` and
  `flickr.photos.recentlyUpdated`.

## Version 0.3.0

- The code has been thoroughly refactored, and fully
  [documented](http://rubydoc.info/github/janko-m/flickr-objects/master/frames). Yay!
- Removed `Flickr.api_methods` – the consequences of its existence made the
  code uglier.
- Added `#find_by` and `#filter` to
  [`Flickr::Object::List`](http://rubydoc.info/github/janko-m/flickr-objects/master/Flickr/Object/List).
- You can now access the raw JSON response of a Flickr object with
  `#attributes`.
- Increased the default timeouts, so now the HTTP requests shouldn't timeout on
  normal internet connections

## Version 0.2.0

- Covered 17 more methods:

```
flickr.people.getPhotosOf
flickr.people.getUploadStatus

flickr.photos.getFavorites
flickr.photos.getNotInSet
flickr.photos.getRecent
flickr.photos.RecentlyUpdated
flickr.photos.getUntagged
flickr.photos.getWithGeoData
flickr.photos.getWithoutGeoData
flickr.photos.setDates
flickr.photos.setMeta
flickr.photos.setPerms
flickr.photos.setSafetyLevel

flickr.photosets.editMeta
flickr.photosets.orderSets
flickr.photosets.reorderPhotos
flickr.photosets.setPrimaryPhoto
```

## Version 0.1.0

- On objects that have an ID, `#==` will now compare their IDs (useful when you
  get the same object from different sources)
- Implemented dynamic finders (`#find_by_<attribute>`) on `Flickr::List`.
- Implemented authentication. Take a look at [this
  wiki](https://github.com/janko-m/flickr-objects/wiki/Authentication).
- Removed videos completely. Too much hacking, because Flickr's API doesn't have
  a good support for them.
- Implemented smart size methods. Take a look at [this wiki](https://github.com/janko-m/flickr-objects/wiki/Sizes)
- Implemented pagination. Supports both Will Paginate and Kaminari. Take a look
  at [this wiki](https://github.com/janko-m/flickr-objects/wiki/Configuration),
  under "Pagination" for how to enable it.


lib/flickr/api/set.rb
module Flickr
  module Api

    class Set < Abstract

      ##
      # @param params [Hash] See documentation below
      # @return [Flickr::Object::Set]
      # @docs [flickr.photosets.create](http://www.flickr.com/services/api/flickr.photosets.create.html)
      #
      def create(params = {})
        response = post "photosets.create", params
        new_object(:Set, response["photoset"])
      end

      ##
      # @param set_ids [String] Comma-delimited list of set IDs (see documentation below)
      # @return [response]
      # @docs [flickr.photosets.orderSets](http://www.flickr.com/services/api/flickr.photosets.orderSets.html)
      #
      def order(set_ids, params = {})
        post "photosets.orderSets", params.merge(photoset_ids: set_ids)
      end

      ##
      # @param set_id [String]
      # @return [response]
      # @docs [flickr.photosets.delete](http://www.flickr.com/services/api/flickr.photosets.delete.html)
      #
      def delete(set_id, params = {})
        post "photosets.delete", params.merge(photoset_id: set_id)
      end

      ##
      # @param set_id [String]
      # @param params [Hash] See documentation below
      # @return [response]
      # @docs [flickr.photosets.editPhotos](http://www.flickr.com/services/api/flickr.photosets.editPhotos.html)
      #
      def edit_photos(set_id, params = {})
        post "photosets.editPhotos", params.merge(photoset_id: set_id)
      end

      ##
      # @param set_id [String]
      # @param params [Hash] See documentation below
      # @return [Flickr::Object::Set]
      # @docs [flickr.photosets.getInfo](http://www.flickr.com/services/api/flickr.photosets.getInfo.html)
      #
      def get_info(set_id, params = {})
        response = get "photosets.getInfo", params.merge(photoset_id: set_id)
        new_object(:Set, response["photoset"])
      end

      ##
      # @param set_id [String]
      # @param params [Hash] See documentation below
      # @option params [Boolean, Array<String>] :sizes For all sizes use `true`, for specific ones
      #   chuck them into an array (e.g. `["Square 75", "Medium 500"]`).
      # @return [Flickr::Object::List<Flickr::Object::Photo>]
      # @docs [flickr.photosets.getPhotos](http://www.flickr.com/services/api/flickr.photosets.getPhotos.html)
      #
      def get_photos(set_id, params = {})
        response = get "photosets.getPhotos", params.merge(photoset_id: set_id)
        new_list(:Photo, response["photoset"]["photo"].map { |h| h.merge("owner" => response["photoset"]["owner"]) }, response["photoset"])
      end

      ##
      # @param set_id [String]
      # @param photo_id [String]
      # @param params [Hash] See documentation below
      # @return [response]
      # @docs [flickr.photosets.addPhoto](http://www.flickr.com/services/api/flickr.photosets.addPhoto.html)
      #
      def add_photo(set_id, photo_id, params = {})
        post "photosets.addPhoto", params.merge(photoset_id: set_id, photo_id: photo_id)
      end

      ##
      # @param set_id [String]
      # @param photo_ids [String] Comma-delimited list of photo IDs (see documentation below)
      # @return [response]
      # @docs [flickr.photosets.removePhotos](http://www.flickr.com/services/api/flickr.photosets.removePhotos.html)
      #
      def remove_photos(set_id, photo_ids, params = {})
        post "photosets.removePhotos", params.merge(photoset_id: set_id, photo_ids: photo_ids)
      end

      ##
      # @param set_id [String]
      # @param photo_id [String]
      # @return [response]
      # @docs [flickr.photosets.removePhoto](http://www.flickr.com/services/api/flickr.photosets.removePhoto.html)
      #
      def remove_photo(set_id, photo_id, params = {})
        post "photosets.removePhoto", params.merge(photoset_id: set_id, photo_id: photo_id)
      end

      ##
      # @param set_id [String]
      # @param params [Hash] See documentation below
      # @return [response]
      # @docs [flickr.photosets.editMeta](http://www.flickr.com/services/api/flickr.photosets.editMeta.html)
      #
      def edit_meta(set_id, params = {})
        post "photosets.editMeta", params.merge(photoset_id: set_id)
      end

      ##
      # @param set_id [String]
      # @param photo_ids [String] Comma-delimited list of photo IDs (see documentation below)
      # @return [response]
      # @docs [flickr.photosets.reorderPhotos](http://www.flickr.com/services/api/flickr.photosets.reorderPhotos.html)
      #
      def reorder_photos(set_id, photo_ids, params = {})
        post "photosets.reorderPhotos", params.merge(photoset_id: set_id, photo_ids: photo_ids)
      end

      ##
      # @param set_id [String]
      # @param photo_id [String]
      # @return [response]
      # @docs [flickr.photosets.setPrimaryPhoto](http://www.flickr.com/services/api/flickr.photosets.setPrimaryPhoto.html)
      #
      def set_primary_photo(set_id, photo_id, params = {})
        post "photosets.setPrimaryPhoto", params.merge(photoset_id: set_id, photo_id: photo_id)
      end

    end

  end
end
